# 02 - Data Preprocessing

## Objective

The purpose of this notebook is to prepare the California CRMLS Sold dataset for machine learning while preventing information from future observations from influencing model development.

The preprocessing workflow includes:

- Loading and combining the monthly CRMLS Sold datasets.
- Restricting the data to California residential single-family properties.
- Removing invalid records, unusable variables, and data leakage features.
- Engineering additional predictive features.
- Adding school district information through a spatial join.
- Creating chronological training, validation, and testing datasets.
- Removing extreme sale-price outliers using thresholds learned from the training data.
- Imputing missing values using statistics learned from the training data.
- Target encoding the high-cardinality school district feature.
- One-hot encoding lower-cardinality categorical variables.
- Standardizing continuous numerical features using training-derived scaling parameters.
- Performing final data integrity checks.
- Exporting the cleaned datasets for model development and evaluation.

The resulting training, validation, and testing datasets will be used for model training, hyperparameter tuning, and final model evaluation throughout the project.

In [261]:
# ============================================
# Import required libraries
# ============================================

from pathlib import Path

import pandas as pd
import numpy as np
import geopandas as gpd

from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import KFold

# Display all columns when inspecting DataFrames
pd.set_option("display.max_columns", None)

## Load the Combined Dataset

The exploratory analysis produced a filtered dataset containing residential single-family property sales.

In this section, we will load that dataset and verify that it has been imported successfully before beginning the preprocessing steps.

## Remove Invalid Sale Records

Before preprocessing, records with missing or non-positive sale prices are removed.

Because ClosePrice is the prediction target, invalid target values cannot be used for supervised machine learning and would negatively affect model training and evaluation.

In [262]:
# ============================================
# Locate all monthly CSV files
# ============================================

data_path = Path("../data/California")

csv_files = sorted(data_path.glob("*.csv"))

print(f"Found {len(csv_files)} CSV files.")

# ============================================
# Read each monthly CSV file
# ============================================

dataframes = []

for file in csv_files:
    print(f"Loading {file.name}...")
    df = pd.read_csv(file, low_memory=False)
    dataframes.append(df)

# Combine all monthly datasets into one DataFrame
homes = pd.concat(dataframes, ignore_index=True)

print("\nAll files successfully combined!")

# ============================================
# Filter to Residential Single-Family properties
# ============================================

homes = homes[
    (homes["PropertyType"] == "Residential") &
    (homes["PropertySubType"] == "SingleFamilyResidence")
].copy()

Found 31 CSV files.
Loading CRMLSSold20220101_20231231_filled.csv...
Loading CRMLSSold202401_filled.csv...
Loading CRMLSSold202402_filled.csv...
Loading CRMLSSold202403_filled.csv...
Loading CRMLSSold202404_filled.csv...
Loading CRMLSSold202405_filled.csv...
Loading CRMLSSold202406_filled.csv...
Loading CRMLSSold202407_filled.csv...
Loading CRMLSSold202408.csv...
Loading CRMLSSold202409.csv...
Loading CRMLSSold202410.csv...
Loading CRMLSSold202411.csv...
Loading CRMLSSold202412.csv...
Loading CRMLSSold202501_filled.csv...
Loading CRMLSSold202502.csv...
Loading CRMLSSold202503.csv...
Loading CRMLSSold202504.csv...
Loading CRMLSSold202505.csv...
Loading CRMLSSold202506.csv...
Loading CRMLSSold202507.csv...
Loading CRMLSSold202508.csv...
Loading CRMLSSold202509.csv...
Loading CRMLSSold202510.csv...
Loading CRMLSSold202511.csv...
Loading CRMLSSold202512.csv...
Loading CRMLSSold202601.csv...
Loading CRMLSSold202602.csv...
Loading CRMLSSold202603.csv...
Loading CRMLSSold202604.csv...
Loading

## Restrict the Dataset to California Properties

The project focuses specifically on predicting residential home sale prices within California.

Although the monthly CRMLS Sold files primarily contain California properties, a small number of records are located outside the state. These observations are removed so that the modeling dataset remains consistent with the geographic scope of the project.

Removing these records is also important before adding the California school district layer later in the preprocessing pipeline, since properties outside California do not fall within the boundaries of the California School District Areas dataset.

In [263]:
# ============================================
# Remove properties outside California
# ============================================

# Count the number of properties that are not located in California.
non_california_count = (
    homes["StateOrProvince"]
    .ne("CA")
    .sum()
)

print(f"Properties outside California: {non_california_count:,}")

# Keep only properties located in California.
homes = homes[
    homes["StateOrProvince"].eq("CA")
].copy()

# Verify that only California properties remain.
print(f"Remaining properties: {len(homes):,}")
print("\nStates remaining:")
print(homes["StateOrProvince"].value_counts())

Properties outside California: 28
Remaining properties: 411,987

States remaining:
StateOrProvince
CA    411987
Name: count, dtype: int64


In [264]:
# ============================================
# Remove invalid sale prices
# ============================================

# Remove records with missing or non-positive sale prices
homes = homes[
    homes["ClosePrice"].notna() &
    (homes["ClosePrice"] > 0)
].copy()

print(f"Rows after removing invalid sale prices: {len(homes):,}")

# Display basic information about the filtered dataset
print(f"\nRows:    {homes.shape[0]:,}")
print(f"Columns: {homes.shape[1]}")

homes.head()

Rows after removing invalid sale prices: 411,984

Rows:    411,984
Columns: 82


,Flooring,ViewYN,WaterfrontYN,BasementYN,PoolPrivateYN,OriginalListPrice,ListingKey,ListAgentEmail,CloseDate,ClosePrice,ListAgentFirstName,ListAgentLastName,Latitude,Longitude,UnparsedAddress,PropertyType,LivingArea,ListPrice,DaysOnMarket,ListOfficeName,BuyerOfficeName,CoListOfficeName,ListAgentFullName,CoListAgentFirstName,CoListAgentLastName,BuyerAgentMlsId,BuyerAgentFirstName,BuyerAgentLastName,FireplacesTotal,AssociationFeeFrequency,AboveGradeFinishedArea,ListingKeyNumeric,MLSAreaMajor,TaxAnnualAmount,CountyOrParish,MlsStatus,ElementarySchool,AttachedGarageYN,ParkingTotal,BuilderName,PropertySubType,LotSizeAcres,SubdivisionName,BuyerOfficeAOR,YearBuilt,BuyerAgencyCompensationType,StreetNumberNumeric,ListingId,BathroomsTotalInteger,City,BuyerAgencyCompensation,TaxYear,BuildingAreaTotal,BedroomsTotal,ContractStatusChangeDate,ElementarySchoolDistrict,CoBuyerAgentFirstName,PurchaseContractDate,ListingContractDate,BelowGradeFinishedArea,BusinessType,StateOrProvince,CoveredSpaces,MiddleOrJuniorSchool,FireplaceYN,Stories,HighSchool,Levels,LotSizeDimensions,LotSizeArea,MainLevelBedrooms,NewConstructionYN,GarageSpaces,HighSchoolDistrict,PostalCode,AssociationFee,LotSizeSquareFeet,MiddleOrJuniorSchoolDistrict,latfilled,lonfilled,BuyerAgentAOR,ListAgentAOR
3,NaN,True,NaN,NaN,False,2499999.0,556365765,bryanmeathe@gmail.com,2022-01-04,2499999.0,Bryan,Meathe,33.147270,-117.340604,295 Chinquapin Ave,Residential,2645.0,2499999.0,37.0,Coldwell Banker Realty,ERA Ranch & Sea Realty,NaN,Bryan Meathe,NaN,NaN,101046,Bill,Kellaway,NaN,Monthly,NaN,556365765.0,92008 - Carlsbad,NaN,San Diego,Closed,NaN,True,2.0,NaN,SingleFamilyResidence,0.3071,NaN,NorthSanDiegoCounty,2016.0,Item1,295.0,NDP2111696,4.0,Carlsbad,2.25,NaN,NaN,4.0,2022-01-04,NaN,NaN,2021-11-19,2021-10-13,NaN,NaN,CA,NaN,NaN,False,NaN,NaN,ThreeOrMore,NaN,13376.0,0.0,NaN,2.0,Carlsbad Unified,92008,140.0,13376.0,NaN,False,False,NaN,NaN
6,NaN,True,NaN,NaN,False,565000.0,556365286,joseph@labelrealtygroup.com,2022-01-10,640000.0,Joseph,Guzman,34.235979,-117.202269,944 Mecham Drive,Residential,2070.0,565000.0,13.0,Label Realty Group,IRN Realty,NaN,Joseph Guzman,NaN,NaN,HAGUIREN,Rene,Aguilar,NaN,NaN,NaN,556365286.0,287 - Arrowhead Area,NaN,San Bernardino,Closed,NaN,True,1.0,NaN,SingleFamilyResidence,0.0780,NaN,TriCounties,2007.0,Item1,944.0,OC21227640,3.0,Lake Arrowhead,2.00,NaN,NaN,3.0,2022-01-10,NaN,NaN,2021-10-26,2021-10-12,NaN,NaN,CA,NaN,NaN,True,NaN,NaN,ThreeOrMore,NaN,3397.0,0.0,False,1.0,Rim of the World,92352,0.0,3397.0,NaN,False,False,NaN,NaN
7,NaN,False,NaN,NaN,False,349999.0,556365284,eaguayo08@gmail.com,2022-03-23,438000.0,ELIAS,AGUAYO,34.114130,-117.442493,16596 Miller Avenue,Residential,1174.0,349999.0,3.0,INTERNATIONAL REAL ESTATE SVCS,INTERNATIONAL REAL ESTATE SVCS,NaN,ELIAS AGUAYO,NaN,NaN,C32048,ELIAS,AGUAYO,NaN,NaN,NaN,556365284.0,264 - Fontana,NaN,San Bernardino,Closed,NaN,True,2.0,NaN,SingleFamilyResidence,0.2273,NaN,CitrusValley,1960.0,Item1,16596.0,CV21227637,2.0,Fontana,2.00,NaN,NaN,3.0,2022-03-23,NaN,NaN,2021-10-16,2021-10-13,NaN,NaN,CA,NaN,NaN,False,1.0,NaN,One,NaN,9900.0,3.0,False,2.0,Fontana Unified,92336,0.0,9900.0,NaN,False,False,NaN,NaN
12,"Carpet,Tile",True,NaN,NaN,True,599000.0,556364261,gcrb714@gmail.com,2022-01-10,615000.0,Rhonda,Belous,33.783767,-116.447280,36353 Paseo Del Sol,Residential,1996.0,599000.0,81.0,"Whitestar Management, Inc",RE/MAX Consultants,"Whitestar Management, Inc",Rhonda Belous,Patrick,Belous,CDAR-43944,Dawn,Jonker,NaN,Monthly,NaN,556364261.0,336 - Cathedral City South,NaN,Riverside,Closed,NaN,False,4.0,NaN,SingleFamilyResidence,0.1400,Rio Del Sol,CaliforniaDesert,2005.0,Item1,36353.0,219068900DA,2.0,Cathedral City,2.00,NaN,NaN,2.0,2022-01-10,NaN,NaN,2022-01-04,2021-10-13,NaN,NaN,CA,NaN,NaN,True,NaN,NaN,NaN,NaN,6098.0,NaN,NaN,2.0,NaN,92234,195.0,6098.0,NaN,False,False,NaN,NaN
14,NaN,True,NaN,NaN,False,399990.0,556363890,allegra@3treehomes.com,2022-01-19,399990.0,Allegra,Gigante-Luft,39.767729,-121.586400,6091 Vista Knolls Drive,Residential,14

## Handle Missing Values

Before preparing the data for machine learning, we will examine the amount of missing data across the dataset.

Missing values can negatively affect model performance, so it is important to determine an appropriate strategy for each variable. Depending on the amount and importance of the missing data, we may:

- Drop observations with missing values.
- Impute missing values using an appropriate statistic.
- Leave variables unchanged if missing values are negligible.

The decisions made in this section will be documented and used consistently throughout the remainder of the project.

In [265]:
# ============================================
# Check missing values across the entire dataset
# ============================================

missing = homes.isnull().sum().to_frame(name="Missing Values")

missing["Percent Missing"] = (
    missing["Missing Values"] / len(homes) * 100
).round(2)

# Display only columns with missing values
missing = missing[missing["Missing Values"] > 0]

# Sort from most missing to least missing
missing = missing.sort_values(
    by="Percent Missing",
    ascending=False
)

missing

,Missing Values,Percent Missing
FireplacesTotal,411984,100.00
TaxAnnualAmount,411984,100.00
AboveGradeFinishedArea,411984,100.00
BusinessType,411984,100.00
CoveredSpaces,411984,100.00
...,...,...
PurchaseContractDate,144,0.03
BathroomsTotalInteger,75,0.02
ListAgentFullName,88,0.02
ListAgentLastName,37,0.01


## Remove Empty Columns

Some variables contain missing values for every observation.

Because these columns contain no usable information, they will be removed before
continuing the preprocessing workflow. The remaining variables will then be
evaluated individually to determine the appropriate strategy for handling any
remaining missing values.

In [266]:
# ============================================
# Remove columns with 100% missing values
# ============================================

# Identify columns where every value is missing
empty_columns = homes.columns[homes.isnull().all()]

print(f"Columns with 100% missing values: {len(empty_columns)}")

# Display their names
for col in empty_columns:
    print(col)

# Remove them
homes = homes.drop(columns=empty_columns)

print(f"\nRemaining columns: {homes.shape[1]}")

Columns with 100% missing values: 8
FireplacesTotal
AboveGradeFinishedArea
TaxAnnualAmount
TaxYear
ElementarySchoolDistrict
BusinessType
CoveredSpaces
MiddleOrJuniorSchoolDistrict

Remaining columns: 74


## Handle Remaining Missing Values

After removing columns with no usable information, we will examine the remaining
variables that still contain missing values.

Rather than applying a single strategy to every variable, we will identify which
columns contain missing data and determine the most appropriate treatment based
on the amount of missing information and the role of each variable in the model.

In [267]:
# ============================================
# Summarize remaining missing values
# ============================================

missing = homes.isnull().sum().to_frame(name="Missing Values")

missing["Percent Missing"] = (
    missing["Missing Values"] / len(homes) * 100
).round(2)

# Keep only columns that still contain missing values
missing = missing[missing["Missing Values"] > 0]

# Sort from highest to lowest percentage missing
missing = missing.sort_values(
    by="Percent Missing",
    ascending=False
)

missing

,Missing Values,Percent Missing
WaterfrontYN,411798,99.95
BelowGradeFinishedArea,409570,99.41
BasementYN,401921,97.56
BuilderName,392657,95.31
LotSizeDimensions,386649,93.85
BuildingAreaTotal,386470,93.81
CoBuyerAgentFirstName,375058,91.04
ElementarySchool,354811,86.12
MiddleOrJuniorSchool,354181,85.97
HighSchool,337727,81.98


## Remove Variables with Excessive Missing Data

Some variables still contain missing values for the majority of observations.

For the baseline model, variables with more than 50% missing values will be removed.
Retaining these variables would require extensive imputation while contributing
relatively little information to the predictive model.

The remaining variables contain sufficient data to support preprocessing and
model development.

In [268]:
# ============================================
# Remove columns with more than 50% missing values
# ============================================

# Remove variables with more than 50% missing values
# to simplify the baseline model.
missing_threshold = 50

columns_to_drop = missing[
    missing["Percent Missing"] > missing_threshold
].index

print(f"Columns removed: {len(columns_to_drop)}")

for column in columns_to_drop:
    print(column)

homes = homes.drop(columns=columns_to_drop)

print(f"\nRemaining columns: {homes.shape[1]}")

Columns removed: 19
WaterfrontYN
BelowGradeFinishedArea
BasementYN
BuilderName
LotSizeDimensions
BuildingAreaTotal
CoBuyerAgentFirstName
ElementarySchool
MiddleOrJuniorSchool
HighSchool
CoListAgentFirstName
CoListAgentLastName
CoListOfficeName
AssociationFeeFrequency
BuyerAgencyCompensation
BuyerAgencyCompensationType
SubdivisionName
latfilled
lonfilled

Remaining columns: 55


## Remove Identifier and Administrative Variables

Some variables uniquely identify a property, listing, office, or agent rather than
describing the property's characteristics.

These variables are removed because they are not expected to improve prediction of
home prices and would unnecessarily increase the complexity of the dataset.

In [269]:
# ============================================
# Remove identifier and administrative columns
# ============================================

columns_to_drop = [
    "ListingId",
    "UnparsedAddress",
    "ListAgentFirstName",
    "ListAgentLastName",
    "ListAgentFullName",
    "ListAgentEmail",
    "BuyerAgentFirstName",
    "BuyerAgentLastName",
    "BuyerAgentMlsId",
    "ListOfficeName",
    "BuyerOfficeName",
]

homes = homes.drop(columns=columns_to_drop)

print(f"Remaining columns: {homes.shape[1]}")

Remaining columns: 44


In [270]:
# ============================================
# Display remaining categorical columns
# ============================================

categorical_columns = homes.select_dtypes(
    include=["object", "string"]
).columns

for column in categorical_columns:
    print(f"{column}: {homes[column].nunique()} unique values")

Flooring: 342 unique values
ViewYN: 2 unique values
PoolPrivateYN: 2 unique values
CloseDate: 1453 unique values
PropertyType: 1 unique values
MLSAreaMajor: 1094 unique values
CountyOrParish: 64 unique values
MlsStatus: 1 unique values
AttachedGarageYN: 2 unique values
PropertySubType: 1 unique values
BuyerOfficeAOR: 74 unique values
City: 1155 unique values
ContractStatusChangeDate: 1450 unique values
PurchaseContractDate: 2056 unique values
ListingContractDate: 2334 unique values
StateOrProvince: 1 unique values
FireplaceYN: 2 unique values
Levels: 18 unique values
NewConstructionYN: 2 unique values
HighSchoolDistrict: 447 unique values
PostalCode: 3729 unique values
BuyerAgentAOR: 63 unique values
ListAgentAOR: 61 unique values


## Remove Constant and Unused Date Variables

Some variables contain only a single unique value after filtering, while others
represent dates that are not required for the baseline predictive model.

These variables are removed to simplify the dataset before encoding the remaining
categorical features.

In [271]:
# ============================================
# Remove constant and unused date columns
# ============================================

columns_to_drop = [
    "PropertyType",
    "PropertySubType",
    "MlsStatus",
    "ContractStatusChangeDate",
    "PurchaseContractDate",
    "ListingContractDate",
]

homes = homes.drop(columns=columns_to_drop)

print(f"Remaining columns: {homes.shape[1]}")

Remaining columns: 38


## Remove High-Cardinality Categorical Variables

Some categorical variables contain hundreds or thousands of unique values.

For the baseline Linear Regression model, these variables are removed to keep the feature space manageable and reduce model complexity.

These variables may be explored in future modeling experiments using alternative encoding strategies.

In [272]:
# ============================================
# Remove high-cardinality / low-value categorical variables
# ============================================

columns_to_drop = [
    "Flooring",
    "MLSAreaMajor",
    "City",
    "HighSchoolDistrict",
    "PostalCode",

    # Little predictive value or extremely high cardinality
    "StateOrProvince",
    "BuyerOfficeAOR",
    "BuyerAgentAOR",
    "ListAgentAOR",
]

homes = homes.drop(columns=columns_to_drop)

print(f"Remaining columns: {homes.shape[1]}")

Remaining columns: 29


## Remove Data Leakage Features

Certain variables contain information that is too closely related to the target
variable (`ClosePrice`) and could artificially inflate model performance.

Following the internship project guidance, `ListPrice` and `OriginalListPrice` are removed
before model development to prevent data leakage. Identifier variables are also
removed because they uniquely identify listings rather than describe the
properties themselves.

## Feature Engineering

Additional predictive variables are created from the original property
characteristics to better capture relationships that may influence sale price.

The engineered features include:

- **PropertyAge** – Age of the home at the time of sale.
- **BathroomBedroomRatio** – Number of bathrooms relative to bedrooms.
- **LivingAreaPerBedroom** – Average interior living space per bedroom.
- **LotSizePerLivingArea** – Relative lot size compared to the home's living area.

These engineered variables are created before the chronological split. Any missing values produced by ratio calculations are left unresolved at this stage and are later imputed using statistics learned only from the training dataset.

In [273]:
# ============================================
# Remove data leakage and identifier variables
# ============================================

columns_to_drop = [
    "ListPrice",
    "OriginalListPrice",
    "DaysOnMarket",
    "ListingKey",
    "ListingKeyNumeric",
]

homes = homes.drop(columns=columns_to_drop)

# ============================================
# Feature Engineering
# ============================================

# Convert CloseDate to datetime.
# The CRMLS files contain multiple date formats,
# so pandas infers the appropriate format for each value.
homes["CloseDate"] = pd.to_datetime(
    homes["CloseDate"],
    format="mixed"
)

print(f"Missing dates after conversion: {homes['CloseDate'].isna().sum()}")

# --------------------------------------------
# Property Age
# --------------------------------------------

# Calculate how old each property was when it sold.
homes["PropertyAge"] = (
    homes["CloseDate"].dt.year - homes["YearBuilt"]
).clip(lower=0)

# --------------------------------------------
# Bathroom-to-Bedroom Ratio
# --------------------------------------------

# Calculate the number of bathrooms relative to bedrooms.
# Bedrooms equal to zero are temporarily replaced with NaN
# to avoid division by zero.
homes["BathroomBedroomRatio"] = (
    homes["BathroomsTotalInteger"] /
    homes["BedroomsTotal"].replace(0, np.nan)
)

# --------------------------------------------
# Living Area Per Bedroom
# --------------------------------------------

# Calculate the average amount of living space
# available per bedroom.
homes["LivingAreaPerBedroom"] = (
    homes["LivingArea"] /
    homes["BedroomsTotal"].replace(0, np.nan)
)

# --------------------------------------------
# Lot Size Per Living Area
# --------------------------------------------

# Compare the size of the property's lot with
# the amount of interior living space.
homes["LotSizePerLivingArea"] = (
    homes["LotSizeSquareFeet"] /
    homes["LivingArea"].replace(0, np.nan)
)

# YearBuilt is no longer needed after creating PropertyAge.
homes = homes.drop(columns="YearBuilt")

print(f"Remaining columns: {homes.shape[1]}")

Missing dates after conversion: 0
Remaining columns: 27


## Add School District Information

Location is one of the strongest predictors of residential property values.

While latitude and longitude provide precise geographic coordinates, they do not directly capture differences in local school systems. To provide a more meaningful geographic feature, each property's coordinates are spatially joined with the California School District Areas (2024–2025) dataset.

The boundary dataset contains Unified, High, and Elementary school districts. Unified districts are retained where available because they represent the complete school system for that area. In locations with separate Elementary and High school districts, the High district is retained so that each property is assigned to a single district rather than appearing multiple times in the dataset.

The resulting school district is added as a categorical feature and will be included alongside the other predictor variables during model training.

In [274]:
# ============================================
# Load California school district boundaries
# ============================================

districts = gpd.read_file(
    "../data/SchoolDistricts/California_School_District_Areas_2024-25.geojson"
)

print(f"School districts loaded: {len(districts):,}")

districts[["DistrictName", "DistrictType"]].head()

School districts loaded: 937


,DistrictName,DistrictType
0,Alameda Unified,Unified
1,Albany City Unified,Unified
2,Berkeley Unified,Unified
3,Castro Valley Unified,Unified
4,Emery Unified,Unified


In [275]:
# ============================================
# Select one school district layer per property
# ============================================

# California contains Unified, High, and Elementary
# school district boundaries.
#
# Unified districts cover the full K-12 school system.
# In areas without a Unified district, High and Elementary
# district boundaries overlap. We retain the High district
# in those areas so each property receives only one
# school district assignment.

districts_model = districts[
    districts["DistrictType"].isin(["Unified", "High"])
].copy()

print(
    f"Districts retained for spatial join: "
    f"{len(districts_model):,}"
)

print("\nDistrict types retained:")
print(
    districts_model["DistrictType"]
    .value_counts()
)

Districts retained for spatial join: 421

District types retained:
DistrictType
Unified    345
High        76
Name: count, dtype: int64


In [276]:
# ============================================
# Convert property coordinates into geometry
# ============================================

homes_geo = gpd.GeoDataFrame(
    homes,
    geometry=gpd.points_from_xy(
        homes["Longitude"],
        homes["Latitude"]
    ),
    crs="EPSG:4326"
)

print(f"Properties converted to points: {len(homes_geo):,}")

Properties converted to points: 411,984


In [277]:
# Match the school district boundary CRS to the
# coordinate reference system used by the properties.
districts_model = districts_model.to_crs(
    homes_geo.crs
)

print(districts_model.crs)

EPSG:4326


In [278]:
# ============================================
# Spatial join with selected school districts
# ============================================

# Match each property's coordinates to either its
# Unified district or, where school systems are split,
# its High school district.
homes_geo = gpd.sjoin(
    homes_geo,
    districts_model[
        ["DistrictName", "DistrictType", "geometry"]
    ],
    how="left",
    predicate="intersects"
)

print(
    homes_geo["DistrictName"]
    .value_counts()
    .head()
)

DistrictName
Los Angeles Unified           39728
San Diego Unified             10916
Antelope Valley Union High     7736
Capistrano Unified             7354
Desert Sands Unified           7244
Name: count, dtype: int64


In [279]:
# ============================================
# Verify the school district spatial join
# ============================================

# Each original property should appear only once
# after the spatial join.
district_matches_per_property = (
    homes_geo.index
    .value_counts()
)

print(f"Rows after spatial join: {len(homes_geo):,}")

print(
    f"Original properties represented: "
    f"{district_matches_per_property.size:,}"
)

print(
    f"Properties matched to multiple districts: "
    f"{(district_matches_per_property > 1).sum():,}"
)

print(
    f"Maximum district matches for one property: "
    f"{district_matches_per_property.max():,}"
)

print(
    f"Properties without a district match: "
    f"{homes_geo['DistrictName'].isna().sum():,}"
)

print("\nMatched district types:")
print(
    homes_geo["DistrictType"]
    .value_counts(dropna=False)
)

Rows after spatial join: 411,984
Original properties represented: 411,984
Properties matched to multiple districts: 0
Maximum district matches for one property: 1
Properties without a district match: 594

Matched district types:
DistrictType
Unified    308267
High       103123
NaN           594
Name: count, dtype: int64


In [280]:
# ============================================
# Return to a regular DataFrame
# ============================================

# Geometry and DistrictType were needed for the
# spatial join but will not be used directly by
# the machine learning models.
homes = homes_geo.drop(
    columns=[
        "geometry",
        "index_right",
        "DistrictType"
    ]
)

print(f"Remaining columns: {homes.shape[1]}")

Remaining columns: 28


## Handle Unmatched School Districts

A small number of properties do not intersect one of the selected Unified or High school district boundaries.

Rather than removing these otherwise valid property sales, unmatched school district values are assigned to an **Unknown** category. This preserves the observations while allowing the school district feature to be used consistently during model development.

In [281]:
# ============================================
# Handle unmatched school districts
# ============================================

# Count properties that did not receive a school
# district during the spatial join.
missing_districts = homes["DistrictName"].isna().sum()

print(
    f"Properties without a school district: "
    f"{missing_districts:,}"
)

# Preserve these property records by assigning
# unmatched districts to an Unknown category.
homes["DistrictName"] = (
    homes["DistrictName"]
    .fillna("Unknown")
)

# Verify that every property now has a district value.
print(
    f"Missing school districts after filling: "
    f"{homes['DistrictName'].isna().sum():,}"
)

print(
    f"School district categories represented: "
    f"{homes['DistrictName'].nunique():,}"
)

print(f"Dataset shape before encoding: {homes.shape}")

Properties without a school district: 594
Missing school districts after filling: 0
School district categories represented: 407
Dataset shape before encoding: (411984, 28)


## Create the Training, Validation, and Testing Sets

To simulate a real-world prediction scenario, the dataset is divided chronologically rather than randomly.

The most recent month of available data is reserved as the **testing set**, while the month immediately before it is reserved as the **validation set**. The 30 months preceding the validation month are used for model training.

Each dataset serves a different purpose:

- **Training set** – Used to fit the machine learning models.
- **Validation set** – Used to compare model configurations and select hyperparameters without using the final testing data.
- **Testing set** – Kept separate until the final model configuration has been selected and is used to measure final predictive performance.

This chronological structure helps prevent future observations from influencing models that are intended to predict later home sales.

## Remove Extreme Sale Price Outliers

Extreme sale prices can disproportionately influence machine learning models, particularly tree-based methods.

The 0.5th and 99.5th percentile sale-price thresholds are computed using the training data only. These same thresholds are then applied to the training, validation, and testing datasets.

Computing the thresholds only from the training period prevents information from the validation or testing periods from influencing preprocessing decisions.

In [282]:
# ============================================
# Create chronological Train/Validation/Test sets
# ============================================

# Identify the most recent month in the dataset.
# This month will be kept as the final testing period.
test_month = homes["CloseDate"].dt.to_period("M").max()

# Use the month immediately before the test month
# as the validation period for model tuning.
validation_month = test_month - 1

# Use the 30 months immediately before the validation
# month as the training period.
training_months = 30

train_start = validation_month - training_months

print(f"Training start month: {train_start}")
print(f"Validation month:     {validation_month}")
print(f"Testing month:        {test_month}")

# --------------------------------------------
# Create the training set
# --------------------------------------------

# Keep observations from the 30-month training window.
# The validation and testing months are excluded.
train_df = homes[
    (homes["CloseDate"].dt.to_period("M") >= train_start) &
    (homes["CloseDate"].dt.to_period("M") < validation_month)
].copy()

# --------------------------------------------
# Create the validation set
# --------------------------------------------

# Keep only observations from the month immediately
# before the testing month.
validation_df = homes[
    homes["CloseDate"].dt.to_period("M") == validation_month
].copy()

# --------------------------------------------
# Create the testing set
# --------------------------------------------

# Keep only observations from the most recent month.
# This dataset will remain untouched during model tuning.
test_df = homes[
    homes["CloseDate"].dt.to_period("M") == test_month
].copy()

print("\nRows before outlier filtering:")
print(f"Training rows:   {len(train_df):,}")
print(f"Validation rows: {len(validation_df):,}")
print(f"Testing rows:    {len(test_df):,}")

# ============================================
# Remove extreme sale-price outliers
# ============================================

# Calculate the lower and upper price thresholds
# using ONLY the training data.
lower_cutoff = train_df["ClosePrice"].quantile(0.005)
upper_cutoff = train_df["ClosePrice"].quantile(0.995)

print(f"\nLower price cutoff: ${lower_cutoff:,.0f}")
print(f"Upper price cutoff: ${upper_cutoff:,.0f}")

# Apply the training-derived thresholds to all three
# datasets so the same price range is used consistently.
train_df = train_df[
    (train_df["ClosePrice"] >= lower_cutoff) &
    (train_df["ClosePrice"] <= upper_cutoff)
].copy()

validation_df = validation_df[
    (validation_df["ClosePrice"] >= lower_cutoff) &
    (validation_df["ClosePrice"] <= upper_cutoff)
].copy()

test_df = test_df[
    (test_df["ClosePrice"] >= lower_cutoff) &
    (test_df["ClosePrice"] <= upper_cutoff)
].copy()

print("\nRows after outlier filtering:")
print(f"Training rows:   {len(train_df):,}")
print(f"Validation rows: {len(validation_df):,}")
print(f"Testing rows:    {len(test_df):,}")

Training start month: 2023-11
Validation month:     2026-05
Testing month:        2026-06

Rows before outlier filtering:
Training rows:   327,643
Validation rows: 12,024
Testing rows:    12,857

Lower price cutoff: $190,000
Upper price cutoff: $8,398,950

Rows after outlier filtering:
Training rows:   324,378
Validation rows: 11,901
Testing rows:    12,710


## Impute Missing Values Using the Training Data

After creating the chronological training, validation, and testing sets, remaining missing values are handled without using information from future observations.

For numerical variables, median values are calculated using only the training dataset. These training-derived medians are then applied to the training, validation, and testing datasets.

For categorical variables, the most frequent value is also determined using only the training dataset and applied consistently across all three datasets.

This approach prevents information from the validation or testing periods from influencing preprocessing decisions.

In [283]:
# ============================================
# Impute missing values using training data only
# ============================================

# Identify numerical predictor columns.
# ClosePrice is excluded because it is the target variable.
numeric_columns = train_df.select_dtypes(
    include=["number"]
).columns.drop("ClosePrice")

# Calculate numerical medians using ONLY
# the training dataset.
training_medians = train_df[numeric_columns].median()

# Apply the same training-derived medians to
# all three chronological datasets.
train_df[numeric_columns] = (
    train_df[numeric_columns]
    .fillna(training_medians)
)

validation_df[numeric_columns] = (
    validation_df[numeric_columns]
    .fillna(training_medians)
)

test_df[numeric_columns] = (
    test_df[numeric_columns]
    .fillna(training_medians)
)

# --------------------------------------------
# Impute categorical variables
# --------------------------------------------

categorical_columns = train_df.select_dtypes(
    include=["object", "string"]
).columns

# CloseDate is a datetime column, so it is not included
# in the categorical columns above.
for column in categorical_columns:

    # Learn the most common category using ONLY
    # the training dataset.
    training_mode = train_df[column].mode()[0]

    # Apply that same value to all three datasets.
    train_df[column] = train_df[column].fillna(
        training_mode
    )

    validation_df[column] = validation_df[column].fillna(
        training_mode
    )

    test_df[column] = test_df[column].fillna(
        training_mode
    )

# Verify that missing predictor values have been handled.
print(
    f"Training missing values: "
    f"{train_df.drop(columns='ClosePrice').isna().sum().sum():,}"
)

print(
    f"Validation missing values: "
    f"{validation_df.drop(columns='ClosePrice').isna().sum().sum():,}"
)

print(
    f"Testing missing values: "
    f"{test_df.drop(columns='ClosePrice').isna().sum().sum():,}"
)

Training missing values: 0
Validation missing values: 0
Testing missing values: 0


## Encode Categorical Variables

Machine learning models require numerical input, so the remaining categorical variables must be converted into numerical features.

Two encoding strategies are used based on the number of categories in each variable:

- **Target encoding** is used for `DistrictName` because it contains hundreds of unique school districts. Each district is represented by the average `ClosePrice` observed for that district in the training data.
- **One-hot encoding** is used for the remaining lower-cardinality categorical variables because they contain relatively few categories and can be represented efficiently using binary indicator columns.

To prevent data leakage, all encoding information is learned from the training dataset only. The resulting mappings and category structure are then applied unchanged to the validation and testing datasets.

In [284]:
# ============================================
# Examine categorical variables
# ============================================

# Display the number of categories in each
# remaining categorical predictor.
categorical_columns = train_df.select_dtypes(
    include=["object", "string"]
).columns

for column in categorical_columns:
    print(
        f"{column}: "
        f"{train_df[column].nunique():,} unique values"
    )

ViewYN: 2 unique values
PoolPrivateYN: 2 unique values
CountyOrParish: 63 unique values
AttachedGarageYN: 2 unique values
FireplaceYN: 2 unique values
Levels: 18 unique values
NewConstructionYN: 2 unique values
DistrictName: 400 unique values


## Target Encode School District

`DistrictName` is a high-cardinality categorical variable, containing hundreds of unique school districts. One-hot encoding every district would create hundreds of additional predictor columns.

Instead, target encoding represents each school district using its average home sale price.

To prevent target leakage within the training data, **5-fold out-of-fold target encoding** is used. For each portion of the training data, district averages are calculated from the other training folds so that a property's own `ClosePrice` does not contribute to its encoded value.

After the training data is encoded, a final district-to-average-price mapping is calculated from the complete training dataset and applied to the validation and testing datasets. Their sale prices are never used to construct the encoded feature.

If a district appears in a held-out fold, validation set, or testing set without an available training-derived district average, the overall mean sale price from the training data is used as a fallback.

In [285]:
# ============================================
# Target encode DistrictName
# ============================================

# Calculate the overall average sale price using
# ONLY the training dataset.
global_mean_price = train_df["ClosePrice"].mean()

print(
    f"Overall training mean ClosePrice: "
    f"${global_mean_price:,.2f}"
)

# ============================================
# Out-of-fold target encoding for training data
# ============================================

# Create an empty Series that will store the
# target-encoded value for every training property.
train_district_encoded = pd.Series(
    index=train_df.index,
    dtype=float
)

# Split the training data into five folds.
#
# For each fold, district averages are calculated
# from the other four folds. This prevents a property's
# own ClosePrice from contributing to its encoded value.
kf = KFold(
    n_splits=5,
    shuffle=True,
    random_state=42
)

for fit_positions, holdout_positions in kf.split(train_df):

    # Select the rows used to calculate district averages.
    fit_fold = train_df.iloc[fit_positions]

    # Select the rows that will receive those averages.
    holdout_fold = train_df.iloc[holdout_positions]

    # Calculate district-level average sale prices using
    # only the fitting portion of this fold.
    fold_district_map = (
        fit_fold
        .groupby("DistrictName")["ClosePrice"]
        .mean()
    )

    # Map the held-out properties to the district averages
    # learned without using their own ClosePrice values.
    encoded_values = (
        holdout_fold["DistrictName"]
        .map(fold_district_map)
        .fillna(global_mean_price)
    )

    # Store the encoded values in their original rows.
    train_district_encoded.loc[
        holdout_fold.index
    ] = encoded_values

# Add the leakage-safe target-encoded feature
# to the training dataset.
train_df["DistrictName_TE"] = train_district_encoded

# ============================================
# Encode validation and testing data
# ============================================

# After creating leakage-safe training values, learn the
# final district mapping from the ENTIRE training dataset.
#
# This mapping can safely be applied to validation and
# testing because their ClosePrice values are not involved.
district_target_map = (
    train_df
    .groupby("DistrictName")["ClosePrice"]
    .mean()
)

print(
    f"Districts learned from training data: "
    f"{len(district_target_map):,}"
)

# Apply the training-derived mapping to validation.
# Districts not observed during training receive the
# overall training mean sale price.
validation_df["DistrictName_TE"] = (
    validation_df["DistrictName"]
    .map(district_target_map)
    .fillna(global_mean_price)
)

# Apply exactly the same mapping to testing.
test_df["DistrictName_TE"] = (
    test_df["DistrictName"]
    .map(district_target_map)
    .fillna(global_mean_price)
)

# Remove the original text-based district variable
# after its numerical representation has been created.
train_df = train_df.drop(columns="DistrictName")
validation_df = validation_df.drop(columns="DistrictName")
test_df = test_df.drop(columns="DistrictName")

Overall training mean ClosePrice: $1,194,332.76
Districts learned from training data: 400


In [286]:
# ============================================
# Verify DistrictName target encoding
# ============================================

print(
    f"Training DistrictName_TE missing values: "
    f"{train_df['DistrictName_TE'].isna().sum():,}"
)

print(
    f"Validation DistrictName_TE missing values: "
    f"{validation_df['DistrictName_TE'].isna().sum():,}"
)

print(
    f"Testing DistrictName_TE missing values: "
    f"{test_df['DistrictName_TE'].isna().sum():,}"
)

print("\nTarget-encoded district summary:")
print(
    train_df["DistrictName_TE"]
    .describe()
)

Training DistrictName_TE missing values: 0
Validation DistrictName_TE missing values: 0
Testing DistrictName_TE missing values: 0

Target-encoded district summary:
count    3.243780e+05
mean     1.194392e+06
std      6.536325e+05
min      2.086667e+05
25%      7.225429e+05
50%      1.007060e+06
75%      1.508009e+06
max      4.630700e+06
Name: DistrictName_TE, dtype: float64


## One-Hot Encode Lower-Cardinality Variables

The remaining categorical variables contain substantially fewer categories than `DistrictName`.

These variables are converted into binary indicator columns using one-hot encoding. The category structure is established from the training data and then applied consistently to the validation and testing datasets.

Using the training dataset as the reference ensures that all three datasets contain the same predictor columns without allowing future data to influence the training feature structure.

In [287]:
# ============================================
# One-hot encode remaining categorical variables
# ============================================

one_hot_columns = [
    "ViewYN",
    "PoolPrivateYN",
    "AttachedGarageYN",
    "FireplaceYN",
    "NewConstructionYN",
    "Levels",
    "CountyOrParish",
]

# One-hot encode the training data.
train_df = pd.get_dummies(
    train_df,
    columns=one_hot_columns,
    drop_first=True,
    dtype=int
)

# One-hot encode validation and testing data.
validation_df = pd.get_dummies(
    validation_df,
    columns=one_hot_columns,
    drop_first=True,
    dtype=int
)

test_df = pd.get_dummies(
    test_df,
    columns=one_hot_columns,
    drop_first=True,
    dtype=int
)

# ============================================
# Match validation/test columns to training
# ============================================

# The training dataset defines the model's feature structure.
# Reindex validation and testing so they contain exactly
# the same columns in exactly the same order.
validation_df = validation_df.reindex(
    columns=train_df.columns,
    fill_value=0
)

test_df = test_df.reindex(
    columns=train_df.columns,
    fill_value=0
)

print(f"Training shape after encoding:   {train_df.shape}")
print(f"Validation shape after encoding: {validation_df.shape}")
print(f"Testing shape after encoding:    {test_df.shape}")

Training shape after encoding:   (324378, 105)
Validation shape after encoding: (11901, 105)
Testing shape after encoding:    (12710, 105)


In [288]:
# Verify that all three datasets have identical columns.
print(
    "Validation columns match training:",
    validation_df.columns.equals(train_df.columns)
)

print(
    "Testing columns match training:",
    test_df.columns.equals(train_df.columns)
)

print(
    "ClosePrice present in all datasets:",
    all(
        "ClosePrice" in df.columns
        for df in [train_df, validation_df, test_df]
    )
)

Validation columns match training: True
Testing columns match training: True
ClosePrice present in all datasets: True


## Normalize Numerical Features

Continuous numerical predictors can have very different scales. For example, living area is measured in square feet while property age is measured in years.

The continuous numerical features are standardized using `StandardScaler`, which transforms the variables so they are measured on comparable scales.

To prevent data leakage, the scaler is fit using **only the training dataset**. The mean and standard deviation learned from the training data are then applied unchanged to the validation and testing datasets.

Binary indicator variables created through one-hot encoding are left unchanged, and `ClosePrice` is not scaled because it is the prediction target.

In [289]:
# ============================================
# Standardize continuous numerical features
# ============================================

continuous_columns = [
    "Latitude",
    "Longitude",
    "LivingArea",
    "ParkingTotal",
    "LotSizeAcres",
    "PropertyAge",
    "BathroomBedroomRatio",
    "LivingAreaPerBedroom",
    "LotSizePerLivingArea",
    "StreetNumberNumeric",
    "BathroomsTotalInteger",
    "BedroomsTotal",
    "Stories",
    "LotSizeArea",
    "MainLevelBedrooms",
    "GarageSpaces",
    "AssociationFee",
    "LotSizeSquareFeet",
    "DistrictName_TE",
]

# Create the scaler.
scaler = StandardScaler()

# Fit the scaler using ONLY the training data
# and transform the training numerical features.
train_df[continuous_columns] = scaler.fit_transform(
    train_df[continuous_columns]
)

# Apply the SAME training-derived scaling parameters
# to the validation data.
validation_df[continuous_columns] = scaler.transform(
    validation_df[continuous_columns]
)

# Apply the SAME training-derived scaling parameters
# to the testing data.
test_df[continuous_columns] = scaler.transform(
    test_df[continuous_columns]
)

print(
    f"Scaled {len(continuous_columns)} "
    f"continuous numerical features."
)

Scaled 19 continuous numerical features.


In [290]:
# ============================================
# Verify numerical scaling
# ============================================

# The scaled training features should have means
# close to 0 and standard deviations close to 1.
scaling_check = pd.DataFrame({
    "Mean": train_df[continuous_columns].mean(),
    "Std Dev": train_df[continuous_columns].std()
}).round(3)

scaling_check

,Mean,Std Dev
Latitude,0.0,1.0
Longitude,0.0,1.0
LivingArea,0.0,1.0
ParkingTotal,0.0,1.0
LotSizeAcres,-0.0,1.0
PropertyAge,0.0,1.0
BathroomBedroomRatio,-0.0,1.0
LivingAreaPerBedroom,-0.0,1.0
LotSizePerLivingArea,0.0,1.0
StreetNumberNumeric,0.0,1.0


## Remove the Date Variable

`CloseDate` was retained throughout preprocessing so that the chronological training, validation, and testing periods could be created correctly.

After the datasets have been separated and all training-derived preprocessing has been completed, the date variable is no longer needed as a model predictor and is removed from all three datasets.

In [291]:
# ============================================
# Remove CloseDate from the model datasets
# ============================================

# CloseDate was needed to create the chronological
# split but will not be used by the models.
train_df = train_df.drop(columns="CloseDate")
validation_df = validation_df.drop(columns="CloseDate")
test_df = test_df.drop(columns="CloseDate")

print(f"Training set:   {train_df.shape}")
print(f"Validation set: {validation_df.shape}")
print(f"Testing set:    {test_df.shape}")

Training set:   (324378, 104)
Validation set: (11901, 104)
Testing set:    (12710, 104)


## Final Data Integrity Checks

Before exporting the processed datasets, several final checks are performed to verify that the data is ready for machine learning.

These checks confirm that:

- No predictor variables contain missing values.
- The training, validation, and testing datasets contain the same columns in the same order.
- All predictor variables are numeric.
- The target variable (`ClosePrice`) is present in all three datasets.
- Known data leakage variables are no longer present.
- No infinite numerical values remain after feature engineering and preprocessing.

Passing these checks helps ensure that differences in future model performance reflect the models themselves rather than inconsistencies in the underlying datasets.

In [292]:
# ============================================
# Final data integrity checks
# ============================================

# Separate the predictor column names from the target.
feature_columns = [
    column
    for column in train_df.columns
    if column != "ClosePrice"
]

# --------------------------------------------
# Check for missing predictor values
# --------------------------------------------

train_missing = (
    train_df[feature_columns]
    .isna()
    .sum()
    .sum()
)

validation_missing = (
    validation_df[feature_columns]
    .isna()
    .sum()
    .sum()
)

test_missing = (
    test_df[feature_columns]
    .isna()
    .sum()
    .sum()
)

print("Missing predictor values:")
print(f"Training:   {train_missing:,}")
print(f"Validation: {validation_missing:,}")
print(f"Testing:    {test_missing:,}")

# --------------------------------------------
# Check that dataset columns match
# --------------------------------------------

print("\nColumn structure:")

print(
    "Validation columns match training:",
    validation_df.columns.equals(train_df.columns)
)

print(
    "Testing columns match training:",
    test_df.columns.equals(train_df.columns)
)

# --------------------------------------------
# Check that the target is present
# --------------------------------------------

print("\nTarget variable:")

print(
    "ClosePrice present in all datasets:",
    all(
        "ClosePrice" in df.columns
        for df in [train_df, validation_df, test_df]
    )
)

# --------------------------------------------
# Check predictor data types
# --------------------------------------------

non_numeric_features = (
    train_df[feature_columns]
    .select_dtypes(exclude=["number"])
    .columns
    .tolist()
)

print("\nNon-numeric predictor columns:")
print(non_numeric_features)

# --------------------------------------------
# Check for known leakage variables
# --------------------------------------------

leakage_columns = [
    "ListPrice",
    "OriginalListPrice",
    "DaysOnMarket",
    "ListingKey",
    "ListingKeyNumeric",
]

remaining_leakage_columns = [
    column
    for column in leakage_columns
    if column in train_df.columns
]

print("\nKnown leakage columns remaining:")
print(remaining_leakage_columns)

# --------------------------------------------
# Check for infinite numerical values
# --------------------------------------------

train_infinite = np.isinf(
    train_df[feature_columns].to_numpy()
).sum()

validation_infinite = np.isinf(
    validation_df[feature_columns].to_numpy()
).sum()

test_infinite = np.isinf(
    test_df[feature_columns].to_numpy()
).sum()

print("\nInfinite predictor values:")
print(f"Training:   {train_infinite:,}")
print(f"Validation: {validation_infinite:,}")
print(f"Testing:    {test_infinite:,}")

# --------------------------------------------
# Display final dimensions
# --------------------------------------------

print("\nFinal dataset dimensions:")
print(f"Training:   {train_df.shape}")
print(f"Validation: {validation_df.shape}")
print(f"Testing:    {test_df.shape}")

print(f"\nPredictor features: {len(feature_columns)}")

Missing predictor values:
Training:   0
Validation: 0
Testing:    0

Column structure:
Validation columns match training: True
Testing columns match training: True

Target variable:
ClosePrice present in all datasets: True

Non-numeric predictor columns:
[]

Known leakage columns remaining:
[]

Infinite predictor values:
Training:   0
Validation: 0
Testing:    0

Final dataset dimensions:
Training:   (324378, 104)
Validation: (11901, 104)
Testing:    (12710, 104)

Predictor features: 103


## Export the Cleaned Datasets

After all preprocessing and data integrity checks are complete, the final training, validation, and testing datasets are saved as separate CSV files.

These datasets serve different purposes during model development:

- **`train_cleaned.csv`** – Used to train the machine learning models.
- **`validation_cleaned.csv`** – Used to compare model configurations and tune hyperparameters.
- **`test_cleaned.csv`** – Reserved for final evaluation after model selection is complete.

Keeping these datasets separate preserves the chronological evaluation structure established during preprocessing.

In [293]:
# ============================================
# Export cleaned train/validation/test datasets
# ============================================

output_path = Path("../data")

# Save each chronological dataset separately.
train_df.to_csv(
    output_path / "train_cleaned.csv",
    index=False
)

validation_df.to_csv(
    output_path / "validation_cleaned.csv",
    index=False
)

test_df.to_csv(
    output_path / "test_cleaned.csv",
    index=False
)

print("Training dataset saved as train_cleaned.csv")
print("Validation dataset saved as validation_cleaned.csv")
print("Testing dataset saved as test_cleaned.csv")

Training dataset saved as train_cleaned.csv
Validation dataset saved as validation_cleaned.csv
Testing dataset saved as test_cleaned.csv


In [294]:
# Verify the final exported dataset dimensions.
print(f"Final dataset shape: {homes.shape}")
print(f"Training set:   {train_df.shape}")
print(f"Validation set: {validation_df.shape}")
print(f"Testing set:    {test_df.shape}")

Final dataset shape: (411984, 28)
Training set:   (324378, 104)
Validation set: (11901, 104)
Testing set:    (12710, 104)


## Summary

In this notebook, the California CRMLS residential single-family dataset was prepared for machine learning by:

- Restricting the dataset to properties located within California.
- Removing records with missing or non-positive sale prices.
- Removing empty columns, variables with excessive missing data, identifier variables, and known data leakage features.
- Engineering additional property characteristics, including property age and several size- and room-based ratio features.
- Spatially joining property coordinates with California school district boundaries.
- Retaining Unified districts and High districts in areas with overlapping Elementary and High district systems to prevent duplicate property records.
- Assigning unmatched school district observations to an `Unknown` category rather than removing otherwise valid sales.
- Creating a chronological split consisting of 30 training months, one validation month, and the most recent month as the final testing period.
- Calculating extreme sale-price thresholds using only the training dataset and applying those same thresholds consistently to all three datasets.
- Imputing missing numerical and categorical values using statistics learned only from the training data.
- Applying 5-fold out-of-fold target encoding to the high-cardinality `DistrictName` feature while using only training-derived mappings for validation and testing data.
- One-hot encoding the remaining lower-cardinality categorical variables.
- Standardizing continuous numerical predictors using scaling parameters learned only from the training dataset.
- Removing `CloseDate` after it was no longer needed for the chronological split.
- Performing final integrity checks to confirm consistent feature structures, numeric predictors, and the absence of missing, infinite, or known leakage values.
- Exporting separate training, validation, and testing datasets for model development and evaluation.

The final datasets contain **103 predictor features**, substantially reducing the feature space created by the previous school-district one-hot encoding approach while preserving geographic information.

The validation dataset will be used for model selection and hyperparameter tuning, while the testing dataset will remain reserved for final performance evaluation.